Multiple Locations on Courses

alanchangxyz opened the pull request:
## Summary
Adds support on calendar events (data, formatting) for multiple buildings if a course lists more than one meeting location
![image](https://github.com/icssc/AntAlmanac/assets/45449494/0767d69d-2b53-41b3-95fd-d8e277a02ac6)
![image](https://github.com/icssc/AntAlmanac/assets/45449494/25c1f86c-c1d0-4804-8062-88fb5f2a2bdb)
_(see Test Plan re: DBH 1100 appearing twice)_

Map code also tweaked to be compatible with multiple locations, but currently only shows one per class (see Future Followup)

## Test Plan
- Mostly verified on frontend with a temporary location duplication line on course store change (hence why DBH 1100 is shown twice in one of the screenshots), removed before commit to maintain proper functionality. Could not find an actual course with 2+ locations so this was used to stub one for testing
- Not completely tested with save/load due to a CORS error with the dev backend on initial load; I only set up for frontend development for this PR

## Issues

Closes #592 

## Future Followup

Might be confusing for users if a class with multiple locations does not distinguish which days it is held at each of those locations
- the map could show markers for all locations, but this PR leaves it to show only a marker for the first location (which could be the wrong meeting location on a given day). would require some sort of validation in the future to make sure it is showing the correct location(s) for the correct day

alanchangxyz committed:
Handle multiple buildings on calendar events/popups + map todo

alanchangxyz committed:
Update building keys for descriptiveness

ap0nia committed:
fix: backend development (#641)

* fix: backend development

* chore: linked sub-projects in root readme

* chore: fix links and add watching to dev

* fix: built the wrong file for lambda handler

Voark committed:
Fix dev server (#642)

* fix dev server

* increase timeout

* add endpoint comment

* change port back to 3000

EricPedley commented on apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:
```diff
@@ -86,7 +86,7 @@ interface CommonCalendarEvent extends Event {
 
 export interface CourseEvent extends CommonCalendarEvent {
-    bldg: string; // E.g., ICS 174, which is actually building + room
+    bldg: string[]; // E.g., ICS 174, which is actually building + room
```
Idea: replace `bldg: string[]` with `locations: Location[]` and add an
```
interface Location {
    building: string
    room: string
}
```
If you agree this is a good idea, please include the change in this PR. Otherwise lmk what problems you think this would introduce.

alanchangxyz replied:
re: showing correct locations for given days - I have no clue how we could verify this programmatically since it could be set arbitrarily by profs/schedules.

+1 re: introducing the location type (will commit that soon) although I think we can also add a `days: string[]` attribute (and potentially an enum too) for if the days of week end up being supplied programmatically later

EricPedley replied:
IDK if we can even scrape that information about which location is for which day. @ecxyzzy do you know?

ecxyzzy replied:
If there's a one location for one set of days (see screenshot), then the first set of days is suppposed to correspond to the first location, the second set to the second location, etc. If not, then it's arbitrary.
![image](https://github.com/icssc/AntAlmanac/assets/89349085/30680a9a-49e0-4081-aca1-ffc82c97d889)
The issue is that in this case, the API supplies two meeting objects in the `meetings` array, whereas in some cases one meeting object has multiple buildings.

alanchangxyz replied:
just curious, what class is this for testing purposes 🤔 i probably should have asked but i assumed (incorrectly) classes were searchable by only building/room

ecxyzzy replied:
The screenshot above corresponds to section code `44100` for 2022 Spring. Another test case that might be useful is `17085` for the same term, since it has two buildings in one meeting object.

EricPedley reviewed (changes requested):
One comment to address. I also have a question about what you wrote in the Future Followup section: how would we know which days correspond to which location?

KevinWu098 committed:
Colors and Courses duplication fix (#635)

EricPedley committed:
change hard-coded color to random choice from palette (#644)

KevinWu098 committed:
Resolved forwardRef warning for ScheduleNameDialog (#639)

KevinWu098 committed:
Section Code Styling (#626)

ecxyzzy committed:
feat: add support for new prereq tree (#643)

* feat: add support for new prereq tree

* refactor: address requested changes

* fix: add dept/courseNumber to CourseInfo

* fix: make value node checking safer

* fix: use department+courseNumber instead of id

For instance, this renders "I&C SCI 33" instead of "I&CSCI33" with the
new API.

KevinWu098 committed:
The Set of usedColors now comes from all courses in the same term (#646)

alanchangxyz committed:
Add Location type and convert meetings/interfaces to use
